rfc/state-model.rst

```diff
@@ -0,0 +1,299 @@
+A state model for Firmware Update
+=================================
+
+The version 0.7 specification defines a state model based around the lifecycle of a firmware image. It looks like this:
+
+.. image:: permitted-state-transitions.svg
+    :width: 75%
+
+A state model based on an image state is informative, but is not the best basis for the Firmware Update API. In the process of updating a single firmware compoment, the current installed firmware is maintained while a new image is be prepared, verified, and tested; prior to committing to the new firmware image.
+
+It is more helpful to track the state of the Firmware store, that contains these images for a component. From the firmware store state, the states of the individual images can be inferred.
+
+
+Firmware Components
+-------------------
+
+Firmware is likely to be split into logical components that apply to a single element inside a device, e.g. the main firmware, the radio module, the secure processing environment, and so on. Firmware components are referred to by a single numerical identifier, unique for each element in the device. Assigning those numbers and sharing them between the Update Client and the Update Service is not specified here and is *IMPDEF*. This is most likely done at build or integration time.
+
+The Firmware update state model described here applies separately to the firmware store for each component.
+
+Analysis
+--------
+
+When the client first runs after a system restart, it needs to query the state of Firmware Update for each firmware component. The response is used to determine if any FWU operations are required.
+
+For a firmware component, depending on the state or progress of a firmware update, there can be one or more firmware images currently in the firmware store:
+
+1. An image that is actively in use by the system.
+2. An image that is being prepared for installation.
+3. A 'backup' of a previously installed image, used to recover if an attempted update fails.
```

Comment:
The original developers of MCUboot had a requirement for a "cheap" rollback image to be available. The idea was that, even after an upgrade is considered successful, the old image is kept around, and not erased. They had an additional command that would then rollback to this image. It is unclear to me if an image that is covered by point 3 becomes covered by point 4 at some point (such as when it is considered that the upgrade has succeeded).

There is an incompatibility with doing this and having anti-rollback protection.

Reply:
This does conflict with a rollback-prevention policy, unless that policy permits what is an effectively extended TRIAL state. The proposed state model can support variation in the `reboot`-when-in-TRIAL policy:
* The one depicted in the table and graphic has the policy that any reboot while in TRIAL results in rollback of the update
* The bootloader could enforce a maximum `reboot` count (>1) before rejecting a TRIAL image
* The bootloader could enforce a maximum time limit before rejecting a TRIAL image on `reboot`

The downside of such a simplified approach is that it can prevent a failed update from being automatically rolled back, leaving a bricked-for-now device.

Having a `restore` operation that can be used in UPDATED state, which reinstalls the backup image, would achieve the use case that you describe. This could be NOT_SUPPORTED for strict anti-rollback policies, or made compatible with various relaxed anti-rollback policies, such as:
* Rollback to the previously installed image is permitted
* Rollback to the previously installed image is permitted for N days after update (or some other constraint)